File: programs/prediction_game/src/instructions/round.rs
use std::cell::{RefMut};

use anchor_lang::prelude::*;

use crate::errors::ErrorCode;

use crate::state::{Round, Game, Crank, get_price, RoundHistory, RoundHistoryItem};

pub fn first_round_shared<'info>(
    game: &mut RefMut<Game>,
    round_key: Pubkey,
    round: &mut RefMut<Round>,
    price_program: &AccountInfo<'info>, 
    price_feed: &AccountInfo<'info>
) -> Result<()> {
    let game_key = game.address;

    require_keys_eq!(game.price_program, price_program.to_account_info().key(), ErrorCode::RoundPriceProgramNotEqual);
    require_keys_eq!(game.price_feed, price_feed.to_account_info().key(), ErrorCode::RoundPriceFeedNotEqual);

    let oracle = game.oracle;

    game.current_round = round_key;
    game.previous_round = round_key;

    round.owner = game.owner.key();
    round.game = game_key;
    round.address = round_key;

    round.round_number = 1_u32; // starting round is 1

    let now = Clock::get()?.unix_timestamp;

    round.round_length = game.round_length;

    round.round_start_time = now;
    round.round_current_time = now;
    round.round_time_difference = 0;

    round.round_predictions_allowed = true;

    let (price, decimals) = get_price(&oracle, price_program, price_feed).unwrap_or((0, 0)); // production & devnet

    round.round_start_price = price;
    round.round_start_price_decimals = decimals;

    round.round_current_price = price;
    round.round_current_price_decimals = decimals;

    round.round_price_difference = 0;
    round.round_price_difference_decimals = 0;

    round.round_end_price = 0;
    round.round_end_price_decimals = 0;


    round.finished = false;
    round.invalid = false;
    round.settled = false;
    round.fee_collected = false;
    round.cranks_paid = false;

    round.total_fee_collected = 0;

    round.total_amount_settled = 0;
    round.total_predictions = 0;
    round.total_predictions_settled = 0;
    
    round.total_unique_crankers = 0;
    round.total_cranks = 0;
    round.total_amount_paid_to_cranks = 0;
    round.total_cranks_paid = 0;

    Ok(())

}

pub fn init_first_round<'info>(ctx: Context<'_, '_, '_, 'info, InitFirstRound<'info>>) -> Result<()> {

    let game = &mut ctx.accounts.game.load_mut()?;
    let round_key = ctx.accounts.round.to_account_info().key();
    let round = &mut ctx.accounts.round.load_init()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;

    require!(first_round_shared(game, round_key, round, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);

    Ok(())
    
}

pub fn init_stuck_first_round<'info>(ctx: Context<'_, '_, '_, 'info, InitStuckFirstRound<'info>>) -> Result<()> {

    let game = &mut ctx.accounts.game.load_mut()?;
    let round_key = ctx.accounts.round.to_account_info().key();
    let round = &mut ctx.accounts.round.load_mut()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;

    require!(first_round_shared(game, round_key, round, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);

    Ok(())
    
}

fn init_round_shared<'info>(
    game: &mut RefMut<Game>,
    current_round_loader: &AccountLoader<'info, Round>,
    next_round: &mut RefMut<Round>,
    next_round_key: Pubkey,
    round_history_loader: &AccountLoader<'info, RoundHistory>, 
    price_program: &AccountInfo<'info>, 
    price_feed: &AccountInfo<'info>
) -> Result<()> {
    let game_key = game.address;

    let current_round_key = current_round_loader.to_account_info().key();
    require_keys_eq!(current_round_key, game.current_round, ErrorCode::RoundKeyNotGameCurrentKey);

    let current_round = current_round_loader.load_mut()?;

    require!(current_round.finished, ErrorCode::RoundNotFinished);
    require!(current_round.fee_collected, ErrorCode::RoundFeeNotCollected);
    require!(current_round.settled, ErrorCode::RoundNotSettled);
    require!(current_round.cranks_paid, ErrorCode::RoundCranksNotPaid);

    require_keys_eq!(game.price_program, price_program.to_account_info().key(), ErrorCode::RoundPriceProgramNotEqual);
    require_keys_eq!(game.price_feed, price_feed.to_account_info().key(), ErrorCode::RoundPriceFeedNotEqual);

    game.current_round = next_round_key;
    game.previous_round = current_round_key;

    if game.total_volume.saturating_add(current_round.total_up_amount.into()).eq(&u128::MAX) {
        let leftover = game.total_volume.saturating_sub(u128::MAX.saturating_sub(current_round.total_up_amount.into()));
        game.total_volume_rollover += 1;
        game.total_volume = leftover;
    } else {
        game.total_volume = game.total_volume.saturating_add(current_round.total_up_amount.into());
    }

    if game.total_volume.saturating_add(current_round.total_down_amount.into()).eq(&u128::MAX) {
        let leftover = game.total_volume.saturating_sub(u128::MAX.saturating_sub(current_round.total_down_amount.into()));
        game.total_volume_rollover += 1;
        game.total_volume = leftover;
    } else {
        game.total_volume = game.total_volume.saturating_add(current_round.total_down_amount.into());
    }

    next_round.owner = game.owner.key();
    next_round.game = game_key;
    next_round.address = next_round_key;

    next_round.round_number = if {current_round.round_number}.eq(&u32::MAX) {
        1
    } else {
        current_round.round_number.saturating_add(1)
    };

    game.round_number = next_round.round_number;

    let now = Clock::get()?.unix_timestamp;

    next_round.round_length = game.round_length;

    next_round.round_start_time = now;
    next_round.round_current_time = now;
    next_round.round_time_difference = 0;

    next_round.round_predictions_allowed = true;

    let (price, decimals) = get_price(&game.oracle, price_program, price_feed).unwrap_or((0, 0));
    // let price = 0;

    next_round.round_start_price = price;
    next_round.round_start_price_decimals = decimals;

    next_round.round_current_price = price;
    next_round.round_current_price_decimals = decimals;

    next_round.round_price_difference = 0;
    next_round.round_price_difference_decimals = 0;

    next_round.finished = false;
    next_round.invalid = false;
    next_round.settled = false;
    next_round.fee_collected = false;
    next_round.cranks_paid = false;

    next_round.round_winning_direction = 0;

    next_round.total_fee_collected = 0;

    next_round.total_up_amount = 0;
    next_round.total_down_amount = 0;

    next_round.total_amount_settled = 0;
    next_round.total_predictions = 0;
    next_round.total_predictions_settled = 0;

    next_round.total_unique_crankers = 0;
    next_round.total_cranks = 0;
    next_round.total_amount_paid_to_cranks = 0;
    next_round.total_cranks_paid = 0;


    let round_history = &mut round_history_loader.load_mut()?;

    // once the round is finished append it to the history
    let next_record_id = round_history.next_record_id();

    
    round_history.append(RoundHistoryItem {
        record_id: next_record_id,
        address: current_round.address,
        round_number: current_round.round_number,
        round_start_time: current_round.round_start_time,
        round_current_time: current_round.round_current_time,
        round_time_difference: current_round.round_time_difference,
        round_start_price: current_round.round_start_price,
        round_start_price_decimals: current_round.round_start_price_decimals,
        round_current_price: current_round.round_current_price,
        round_current_price_decimals: current_round.round_current_price_decimals,
        round_end_price: current_round.round_end_price,
        round_end_price_decimals: current_round.round_end_price_decimals,
        round_price_difference: current_round.round_price_difference,
        round_price_difference_decimals: current_round.round_price_difference_decimals,
        round_winning_direction: current_round.round_winning_direction,
        round_invalid: current_round.invalid,
        total_fee_collected: current_round.total_fee_collected,
        total_up_amount: current_round.total_up_amount,
        total_down_amount: current_round.total_down_amount,
        total_amount_settled: current_round.total_amount_settled,
        total_predictions_settled: current_round.total_predictions_settled,
        total_predictions: current_round.total_predictions,
        total_unique_crankers: current_round.total_unique_crankers,
        total_cranks: current_round.total_cranks,
        total_cranks_paid: current_round.total_cranks_paid,
        total_amount_paid_to_cranks: current_round.total_amount_paid_to_cranks
    });

    Ok(())
}



pub fn init_second_round<'info>(ctx: Context<'_, '_, '_, 'info, InitSecondRound<'info>>, _next_round_number: [u8; 4]) -> Result<()> {
    let game = &mut ctx.accounts.game.load_mut()?;
    let current_round_loader = &ctx.accounts.first_round;
    let next_round_key = ctx.accounts.second_round.to_account_info().key();
    let next_round = &mut ctx.accounts.second_round.load_init()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;
    let round_history_loader = &ctx.accounts.round_history;

    require!(init_round_shared(game, current_round_loader, next_round, next_round_key, round_history_loader, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);

    Ok(())
}

pub fn init_stuck_second_round<'info>(ctx: Context<'_, '_, '_, 'info, InitStuckSecondRound<'info>>, _next_round_number: [u8; 4]) -> Result<()> {
    let game = &mut ctx.accounts.game.load_mut()?;
    let current_round_loader = &ctx.accounts.first_round;
    let next_round_key = ctx.accounts.second_round.to_account_info().key();
    let next_round = &mut ctx.accounts.second_round.load_mut()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;
    let round_history_loader = &ctx.accounts.round_history;

    require!(init_round_shared(game, current_round_loader, next_round, next_round_key, round_history_loader, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);
    
    Ok(())
}

pub fn init_next_round_and_close_previous<'info>(ctx: Context<'_, '_, '_, 'info, InitNextRoundAndClosePrevious<'info>>, _next_round_number: [u8; 4]) -> Result<()> {
    let game = &mut ctx.accounts.game.load_mut()?;
    let current_round_loader = &ctx.accounts.current_round;
    let next_round_key = ctx.accounts.next_round.to_account_info().key();
    let next_round = &mut ctx.accounts.next_round.load_init()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;
    let round_history_loader = &ctx.accounts.round_history;

    require!(init_round_shared(game, current_round_loader, next_round, next_round_key, round_history_loader, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);

    Ok(())
}

pub fn init_stuck_next_round_and_close_previous<'info>(ctx: Context<'_, '_, '_, 'info, InitStuckNextRoundAndClosePrevious<'info>>, _next_round_number: [u8; 4]) -> Result<()> {
    let game = &mut ctx.accounts.game.load_mut()?;
    let current_round_loader = &ctx.accounts.current_round;
    let next_round_key = ctx.accounts.next_round.to_account_info().key();
    let next_round = &mut ctx.accounts.next_round.load_mut()?;
    let price_program = &ctx.accounts.price_program;
    let price_feed = &ctx.accounts.price_feed;
    let round_history_loader = &ctx.accounts.round_history;

    require!(init_round_shared(game, current_round_loader, next_round, next_round_key, round_history_loader, price_program, price_feed).is_ok(), ErrorCode::FailedToInitRound);

    Ok(())
}


#[derive(Accounts)]
pub struct CloseRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub receiver: SystemAccount<'info>,

    #[account(
        mut,
        constraint = round.load_mut()?.owner == signer.key() @ ErrorCode::SignerNotOwner,
        close = receiver
    )]
    pub round: AccountLoader<'info, Round>
}

#[derive(Accounts)]
pub struct InitFirstRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,

    #[account(
        init,
        seeds = [
            env!("CARGO_PKG_VERSION").as_bytes(), 
            game.key().as_ref(), 
            &[(1_u32).to_be_bytes()[0]], 
            &[(1_u32).to_be_bytes()[1]], 
            &[(1_u32).to_be_bytes()[2]], 
            &[(1_u32).to_be_bytes()[3]], 
            b"round"
        ], 
        bump, 
        payer = signer,
        space = std::mem::size_of::<Round>() + 8
    )]
    pub round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(next_round_number: [u8; 4])]
pub struct InitStuckFirstRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub round_history: AccountLoader<'info, RoundHistory>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,

    #[account(mut)]
    pub round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}


#[derive(Accounts)]
#[instruction(next_round_number: [u8; 4])]
pub struct InitSecondRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub round_history: AccountLoader<'info, RoundHistory>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,

    #[account(
        init,
        seeds = [
            env!("CARGO_PKG_VERSION").as_bytes(), 
            game.key().as_ref(), 
            &[next_round_number[0]], 
            &[next_round_number[1]], 
            &[next_round_number[2]],  
            &[next_round_number[3]],  
            b"round"
        ], 
        bump, 
        payer = signer,
        space = std::mem::size_of::<Round>() + 8
    )]
    pub second_round: AccountLoader<'info, Round>,

    #[account(mut)]
    pub first_round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}


#[derive(Accounts)]
#[instruction(next_round_number: [u8; 4])]
pub struct InitStuckSecondRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub round_history: AccountLoader<'info, RoundHistory>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,

    #[account(mut)]
    pub second_round: AccountLoader<'info, Round>,

    #[account(mut)]
    pub first_round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(next_round_number: [u8; 4])]
pub struct InitNextRoundAndClosePrevious<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub round_history: AccountLoader<'info, RoundHistory>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,
    
    #[account(
        mut,
        constraint = signer.key() == receiver.key()
    )]
    pub receiver: SystemAccount<'info>,

    #[account(
        init,
        seeds = [
            env!("CARGO_PKG_VERSION").as_bytes(), 
            game.key().as_ref(),
            &[next_round_number[0]],
            &[next_round_number[1]],
            &[next_round_number[2]],
            &[next_round_number[3]],
            b"round"
            
        ], 
        bump, 
        payer = signer,
        space = std::mem::size_of::<Round>() + 8
    )]
    pub next_round: AccountLoader<'info, Round>,

    #[account(
        mut
    )]
    pub current_round: AccountLoader<'info, Round>,

    #[account(
        mut,
        close = receiver
    )]
    pub previous_round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(next_round_number: [u8; 4])]
pub struct InitStuckNextRoundAndClosePrevious<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut)]
    pub game: AccountLoader<'info, Game>,

    #[account(mut)]
    pub round_history: AccountLoader<'info, RoundHistory>,

    #[account(mut)]
    pub crank: AccountLoader<'info, Crank>,
    
    #[account(
        mut,
        constraint = signer.key() == receiver.key()
    )]
    pub receiver: SystemAccount<'info>,

    #[account(mut)]
    pub next_round: AccountLoader<'info, Round>,

    #[account(
        mut
    )]
    pub current_round: AccountLoader<'info, Round>,

    #[account(
        mut,
        close = receiver
    )]
    pub previous_round: AccountLoader<'info, Round>,

    /// CHECK:
    #[account()]
    pub price_program: AccountInfo<'info>,

    /// CHECK:
    #[account()]
    pub price_feed: AccountInfo<'info>,

    // required for Account
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct AdminCloseRound<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(mut, close = receiver)]
    pub round: AccountLoader<'info, Round>,
    
    #[account(
        mut
    )]
    pub receiver: SystemAccount<'info>,
}

File: sdk/src/accounts/round.ts
import * as anchor from "@project-serum/anchor"
import { PublicKey } from '@solana/web3.js'
import { Workspace } from '../workspace'
import Game from "./game"
import { U32MAX } from "../constants"
import { DataUpdatable } from "../dataUpdatable"
import { confirmTxRetry } from "../util/index"
import Crank from "./crank"
import { Oracle } from "../types"


export type RoundAccount = {

    owner: PublicKey
    game: PublicKey
    address: PublicKey

    roundNumber: number
    roundLength: anchor.BN
    
    finished: boolean
    invalid: boolean
    settled: boolean
    feeCollected: boolean
    cranksPaid: boolean
 
    roundPredictionsAllowed: boolean,

    roundStartTime: anchor.BN
    roundCurrentTime: anchor.BN
    roundTimeDifference: anchor.BN

    roundStartPrice: anchor.BN
    roundStartPriceDecimals: anchor.BN
    
    roundCurrentPrice: anchor.BN
    roundCurrentPriceDecimals: anchor.BN

    roundEndPrice: anchor.BN
    roundEndPriceDecimals: anchor.BN

    roundPriceDifference: anchor.BN,
    roundPriceDifferenceDecimals: anchor.BN,

    roundWinningDirection: number

    totalFeeCollected: anchor.BN

    totalUpAmount: anchor.BN
    totalDownAmount: anchor.BN

    totalAmountSettled: anchor.BN
    totalPredictionsSettled: number
    totalPredictions: number

    totalUniqueCrankers: number
    totalCranks: number
    totalCranksPaid: number
    totalAmountPaidToCranks: anchor.BN,

    padding01: PublicKey[]

}

export default class Round implements DataUpdatable<RoundAccount> {
    account: RoundAccount

    constructor(account: RoundAccount) {
        this.account = account;
    }

    public async updateData(data: RoundAccount): Promise<boolean> {
        this.account = data;
        return true;
    }

    public static fromJSON<RoundAccount>(json: any): RoundAccount {
        
        let owner = new PublicKey(json.owner);
        let game = new PublicKey(json.game);
        let address = new PublicKey(json.address);
        let roundNumber = json.roundNumber;
        let roundLength = new anchor.BN(json.roundLength, 16);
        let finished = json.finished;
        let invalid = json.invalid;
        let settled = json.settled;
        let feeCollected = json.feeCollected;
        let cranksPaid = json.cranksPaid;
        let roundPredictionsAllowed = json.roundPredictionsAllowed;
        let roundStartTime = new anchor.BN(json.roundStartTime, 16);
        let roundCurrentTime = new anchor.BN(json.roundCurrentTime, 16);
        let roundTimeDifference = new anchor.BN(json.roundTimeDifference, 16);
        let roundStartPrice = new anchor.BN(json.roundStartPrice, 16);
        let roundStartPriceDecimals = new anchor.BN(json.roundStartPriceDecimals, 16);
        let roundCurrentPrice = new anchor.BN(json.roundCurrentPrice, 16);
        let roundCurrentPriceDecimals = new anchor.BN(json.roundCurrentPriceDecimals, 16);
        let roundEndPrice = new anchor.BN(json.roundEndPrice, 16);
        let roundEndPriceDecimals = new anchor.BN(json.roundEndPriceDecimals, 16);
        let roundPriceDifference = new anchor.BN(json.roundPriceDifference, 16);
        let roundPriceDifferenceDecimals = new anchor.BN(json.roundPriceDifferenceDecimals, 16);
        let roundWinningDirection = json.roundWinningDirection;
        let totalFeeCollected = new anchor.BN(json.totalFeeCollected, 16);
        let totalUpAmount = new anchor.BN(json.totalUpAmount, 16);
        let totalDownAmount = new anchor.BN(json.totalDownAmount, 16);
        let totalAmountSettled = new anchor.BN(json.totalAmountSettled, 16);
        let totalPredictionsSettled = json.totalPredictionsSettled;
        let totalPredictions = json.totalPredictions;
        let totalUniqueCrankers = json.totalUniqueCrankers;
        let totalCranks = json.totalCranks;
        let totalCranksPaid = json.totalCranksPaid;
        let totalAmountPaidToCranks = new anchor.BN(json.totalAmountPaidToCranks, 16);
        let padding01 = json.padding01.map((x: string) => new PublicKey(x)) as PublicKey[];

        return {
            owner,
            game,
            address,
            roundNumber,
            roundLength,
            finished,
            invalid,
            settled,
            feeCollected,
            cranksPaid,
            roundPredictionsAllowed,
            roundStartTime,
            roundCurrentTime,
            roundTimeDifference,
            roundStartPrice,
            roundStartPriceDecimals,
            roundCurrentPrice,
            roundCurrentPriceDecimals,
            roundEndPrice,
            roundEndPriceDecimals,
            roundPriceDifference,
            roundPriceDifferenceDecimals,
            roundWinningDirection,
            totalFeeCollected,
            totalUpAmount,
            totalDownAmount,
            totalAmountSettled,
            totalPredictionsSettled,
            totalPredictions,
            totalUniqueCrankers,
            totalCranks,
            totalCranksPaid,
            totalAmountPaidToCranks,
            padding01
        } as unknown as RoundAccount;
    }

    public convertOraclePriceToNumber(price: anchor.BN, decimals_: anchor.BN, game: Game) : number {
        try {
            let decimals = decimals_.abs();
            if (game.account.oracle === Oracle.Chainlink) {
                let scaled_val = price.toString();
                if (scaled_val.length <= decimals.toNumber()) {
                    let zeros = "";
                    for(let x = 0; x < decimals.toNumber() - scaled_val.length; x++) {
                        zeros += "0";
                    }
                    let charArray = [...scaled_val];
                    charArray.splice(0, 0, ...zeros)
                    scaled_val = "0." + charArray.join("")
                    return parseFloat(scaled_val);
                } else {
                    let charArray = Array.from(scaled_val);
                    charArray.splice(charArray.length - decimals.toNumber(), 0, ".")
                    return parseFloat(charArray.join(""))
                }
            } else if (game.account.oracle === Oracle.Pyth || game.account.oracle === Oracle.Switchboard) {
                
                return parseFloat(
                    (price.div((new anchor.BN(10)).pow(decimals)).toNumber() + (price.mod((new anchor.BN(10)).pow(decimals)).toNumber() / (10 ** decimals.toNumber()))).toFixed(2)
                )
            }
        } catch(error) {
            console.error(error);
        }
        
        
    }

    public static initializeFirst(workspace: Workspace, game: Game, crank: Crank): Promise<Game> {
    
        return new Promise((resolve, reject) => {
            workspace.programAddresses.getRoundPubkey(game.account.address, new anchor.BN(1)).then(([roundPubkey, _roundPubkeyBump]) => {
                workspace.program.methods.initFirstRoundInstruction().accounts({
                    signer: workspace.owner,
                    game: game.account.address,
                    crank: crank.account.address,
                    round: roundPubkey,
                    priceProgram: game.account.priceProgram,
                    priceFeed: game.account.priceFeed,
                    systemProgram: anchor.web3.SystemProgram.programId
                }).transaction().then(tx => {
                    workspace.sendTransaction(tx).then(txSignature => {
                        confirmTxRetry(workspace, txSignature).then(() => {
                            game.updateGameData(workspace).then((game: Game) => {
                                game.loadRoundData(workspace).then((game: Game) => {
                                    resolve(game);
                                }).catch(error => {
                                    reject(error);
                                })
                            }).catch(error => {
                                reject(error);
                            })
                        }).catch(error => {
                            reject(error);
                        })
                    }).catch(error => {
                        reject(error);
                    })
                }).catch(error => {
                    reject(error);
                })
            }).catch(error => {
                reject(error);
            })
        })
    }

    public static initializeSecond(workspace: Workspace, game: Game, crank: Crank): Promise<Game> {
        
        return new Promise((resolve, reject) => {
            
            workspace.programAddresses.getRoundPubkey(game.account.address, new anchor.BN(2)).then(([roundPubkey, _secondRoundPubkeyBump]) => {
                let roundNumberAsBuffer = workspace.programAddresses.roundToBuffer(new anchor.BN(2));
                workspace.program.methods.initSecondRoundInstruction([roundNumberAsBuffer[0], roundNumberAsBuffer[1], roundNumberAsBuffer[2], roundNumberAsBuffer[3]]).accounts({
                    signer: workspace.owner,
                    game: game.account.address,
                    crank: crank.account.address,
                    secondRound: roundPubkey,
                    roundHistory: game.account.roundHistory,
                    firstRound: game.currentRound.account.address,
                    priceProgram: game.account.priceProgram,
                    priceFeed: game.account.priceFeed,
                    systemProgram: anchor.web3.SystemProgram.programId
                }).transaction().then(tx => {
                    workspace.sendTransaction(tx).then(txSignature => {
                        confirmTxRetry(workspace, txSignature).then(() => {
                            game.updateGameData(workspace).then((game: Game) => {
                                game.updateRoundData(workspace).then((game: Game) => {
                                    resolve(game);
                                }).catch(error => {
                                    reject(error);
                                })
                            }).catch(error => {
                                reject(error);
                            })
                        }).catch(error => {
                            reject(error);
                        })
                    }).catch(error => {
                        reject(error);
                    })
                }).catch(error => {
                    reject(error);
                })
            }).catch(error => {
                reject(error);
            })
        })
       
    }


    public static initializeNext(workspace: Workspace, game: Game, crank: Crank): Promise<Game> {
        
        return new Promise((resolve, reject) => {
            let nextRoundNumber = new anchor.BN(game.currentRound.account.roundNumber + 1);
            if (nextRoundNumber.gt(U32MAX)) {
                nextRoundNumber = new anchor.BN(1);
            }
            let roundNumberAsBuffer = workspace.programAddresses.roundToBuffer(nextRoundNumber);
            workspace.programAddresses.getRoundPubkey(game.account.address, nextRoundNumber).then(([roundPubkey, _roundPubkeyBump]) => {
                workspace.program.methods.initNextRoundAndClosePreviousInstruction([roundNumberAsBuffer[0], roundNumberAsBuffer[1], roundNumberAsBuffer[2], roundNumberAsBuffer[3]]).accounts({
                    signer: workspace.owner,
                    game: game.account.address,
                    crank: crank.account.address,
                    receiver: workspace.owner,
                    nextRound: roundPubkey,
                    roundHistory: game.account.roundHistory,
                    currentRound: game.currentRound.account.address,
                    previousRound: game.previousRound.account.address,
                    priceProgram: game.account.priceProgram,
                    priceFeed: game.account.priceFeed,
                    systemProgram: anchor.web3.SystemProgram.programId
                }).transaction().then(tx => {
                    workspace.sendTransaction(tx).then(txSignature => {
                        confirmTxRetry(workspace, txSignature).then(() => {
                            game.updateGameData(workspace).then((game: Game) => {
                                game.updateRoundData(workspace).then((game: Game) => {
                                    resolve(game);
                                }).catch(error => {
                                    reject(error);
                                })
                            }).catch(error => {
                                reject(error);
                            })
                        }).catch(error => {
                            reject(error);
                        })
                    }).catch(error => {
                        reject(error);
                    })
                }).catch(error => {
                    reject(error);
                })
            }).catch(error => {
                reject(error);
            })
        })
    }

    public static adminCloseRound(workspace: Workspace, round: Round): Promise<void> {
        
        return new Promise((resolve, reject) => {
            workspace.program.methods.adminCloseRoundInstruction().accounts({
                signer: workspace.owner,
                round: round.account.address,
                receiver: workspace.owner
            }).transaction().then(tx => {
                workspace.sendTransaction(tx).then(txSignature => {
                    confirmTxRetry(workspace, txSignature).then(() => {
                        resolve();
                    }).catch(error => {
                        reject(error);
                    })
                }).catch(error => {
                    reject(error);
                })
            }).catch(error => {
                reject(error);
            })
        })
    }
}